### PR TITLE
reverse search fix

### LINKED ISSUE #39401: Random errors when searching REPL's history with "Ctrl-r"
When I search the history (press Ctrl-r and begin something), I'll occasionally get these errors:
```
│    BoundsError: attempt to access 0-element Vector{UInt8} at index [0]
│    Stacktrace:
│      [1] _rsearchindex(s::Vector{UInt8}, t::Vector{UInt8}, _k::Int64)
│        @ Base ./strings/search.jl:461
│      [2] _rsearchindex(s::String, t::String, i::Int64)
│        @ Base ./strings/search.jl:446
│      [3] _rsearch
│        @ ./strings/search.jl:523 [inlined]
│      [4] findprev
│        @ ./strings/search.jl:556 [inlined]
│      [5] findlast
│        @ ./strings/search.jl:363 [inlined]
│      [6] history_search(hist::REPL.REPLHistoryProvider, query_buffer::IOBuffer, response_buffer::IOBuffer, backwards::Bool, skip_current::Bool)
│        @ REPL /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.7/REPL/src/REPL.jl:744
│      [7] update_display_buffer(s::REPL.LineEdit.SearchState, data::REPL.LineEdit.ModeState)
│        @ REPL.LineEdit /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.7/REPL/src/LineEdit.jl:1689
│      [8] update_display_buffer(s::REPL.LineEdit.MIState, args::Any)
│        @ REPL.LineEdit /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.7/REPL/src/LineEdit.jl:227
│      [9] (::REPL.LineEdit.var"#73#104")(s::REPL.LineEdit.MIState, data::REPL.LineEdit.ModeState, c::Union{Char, SubString{String}, String})
│        @ REPL.LineEdit /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.7/REPL/src/LineEdit.jl:2004
│     [10] #invokelatest#2
│        @ ./essentials.jl:710 [inlined]
│     [11] invokelatest
│        @ ./essentials.jl:708 [inlined]
│     [12] (::REPL.LineEdit.var"#22#23"{REPL.LineEdit.var"#73#104", String})(s::Any, p::Any)
│        @ REPL.LineEdit /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.7/REPL/src/LineEdit.jl:1414
│     [13] prompt!(term::REPL.Terminals.TextTerminal, prompt::REPL.LineEdit.ModalInterface, s::REPL.LineEdit.MIState)
│        @ REPL.LineEdit /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.7/REPL/src/LineEdit.jl:2534
│     [14] run_interface(terminal::REPL.Terminals.TextTerminal, m::REPL.LineEdit.ModalInterface, s::REPL.LineEdit.MIState)
│        @ REPL.LineEdit /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.7/REPL/src/LineEdit.jl:2436
│     [15] run_frontend(repl::REPL.LineEditREPL, backend::REPL.REPLBackendRef)
│        @ REPL /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.7/REPL/src/REPL.jl:1126
│     [16] (::REPL.var"#44#49"{REPL.LineEditREPL, REPL.REPLBackendRef})()
│        @ REPL ./task.jl:406
```
```
julia> versioninfo()
Julia Version 1.7.0-DEV.383
Commit bf15b706e9* (2021-01-26 04:39 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: AMD Ryzen 7 4700U with Radeon Graphics
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-11.0.1 (ORCJIT, znver2)

```

### PR DESCRIPTION
**Pull request by Moelf**
fix #39401, #40244﻿

### CODE REVIEW

**Review by stevengj — DISMISSED** [review]

File: base/strings/search.jl [hunk 1]
```diff
@@ -442,6 +442,8 @@ function _rsearchindex(s::String, t::String, i::Integer)
     # Check for fast case of a single byte
     if lastindex(t) == 1
         return something(findprev(isequal(t[1]), s, i), 0)
+    elseif sizeof(t) > sizeof(s)
+        return 0
     elseif lastindex(t) != 0
         j = i ≤ ncodeunits(s) ? nextind(s, i)-1 : i
         return _rsearchindex(unsafe_wrap(Vector{UInt8}, s), unsafe_wrap(Vector{UInt8}, t), j)
```
Comment (stevengj) [inline base/strings/search.jl:449]: It seems like `_rsearchindex` should be fixed for byte arrays, not just for `String`.  i.e. you are fixing the wrong method.

Comment (Moelf) [reply base/strings/search.jl:449]: It seems to me the byte array code should error and we can't fix it there without substantial re-factoring.

The thing looks to me people expect `find(string...)` to return `nothing` when the same situation should `error`:

```
julia> findlast("xxx", "")

julia> findlast(UInt8[1,2,3], UInt8[])
ERROR: BoundsError: attempt to access 0-element Vector{UInt8} at index [0]
```

I personally think the first one should error too, but 1.5 didn't have that behavior thus we're fixing this bug

Comment (Moelf) [reply base/strings/search.jl:449]: https://github.com/JuliaLang/julia/blob/bc3ce48636b700d153d2c84614aaea8cb8843d43/base/strings/search.jl#L462

simply changing this to `k<0` (without any other changes in this PR) seems to pass all the test (old + new), what you think? @stevengj

File: test/strings/search.jl [hunk 2]
```diff
@@ -97,6 +97,7 @@ for str in [astr]
     @test findprev('l', str, 2) == nothing
     @test findlast(',', str) == 6
     @test findprev(',', str, 5) == nothing
+    @test findlast(str, "") == nothing
```
Comment (stevengj) [inline test/strings/search.jl:100]: Might also be good to add
```jl
@test findlast(str^2, str) == nothing
```
i.e. a case where the pattern is longer than the string to be searched, but the latter is nonempty.

**Review by stevengj — APPROVED** [review]

**Pull request merged**